### PR TITLE
executor: Add replace support for LOAD DATA

### PR DESCRIPTION
**Pull request by dveeden**
### What problem does this PR solve?

Issue Number: ref #24515

Problem Summary:

This adds support for `LOAD DATA... REPLACE INTO...`

### What is changed and how it works?

```
mysql> TABLE t2;
+----+----------+
| id | c1       |
+----+----------+
|  1 | test     |
|  2 | test two |
+----+----------+
2 rows in set (0.00 sec)

mysql> LOAD DATA LOCAL INFILE '/tmp/t2.csv' INTO TABLE t2;
Query OK, 1 row affected, 2 warnings (0.00 sec)
Records: 3  Deleted: 0  Skipped: 2  Warnings: 2

mysql> \! cat /tmp/t2.csv
1       XXXtest
2       XXXtest two
3       XXXtest three
mysql> TABLE t2;
+----+---------------+
| id | c1            |
+----+---------------+
|  1 | test          |
|  2 | test two      |
|  3 | XXXtest three |
+----+---------------+
3 rows in set (0.00 sec)

mysql> LOAD DATA LOCAL INFILE '/tmp/t2.csv' REPLACE INTO TABLE t2;
Query OK, 3 rows affected (0.00 sec)
Records: 3  Deleted: 0  Skipped: 0  Warnings: 0

mysql> TABLE t2;
+----+---------------+
| id | c1            |
+----+---------------+
|  1 | XXXtest       |
|  2 | XXXtest two   |
|  3 | XXXtest three |
+----+---------------+
3 rows in set (0.00 sec)
```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [x] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

### Release note

```release-note
The REPLACE keyword for `LOAD DATA LOCAL INFILE` is now supported
```

### TODO

- [x] Check permissions (permission to delete rows, etc)
- [x] Check statistics (`Deleted: 0` might be wrong?)
- [x] Check if the old and new row are identical and not replace if that's the case
- [x] Add tests

### CODE REVIEW

**Review by morgo — APPROVED** [review]
LGTM but please have someone on runtime review as well.

cc @XuHuaiyu

File: executor/insert_common.go [hunk 1]
```diff
@@ -1087,10 +1091,16 @@ func (e *InsertValues) batchCheckAndInsert(ctx context.Context, rows [][]types.D
...
 			_, err := txn.Get(ctx, r.handleKey.newKey)
 			if err == nil {
-				e.ctx.GetSessionVars().StmtCtx.AppendWarning(r.handleKey.dupErr)
-				continue
-			}
-			if !kv.IsErrNotFound(err) {
+				if replace {
+					err2 := e.removeRow(ctx, txn, r)
+					if err2 != nil {
+						return err2
```
Comment (tiancaiamao) [inline executor/insert_common.go:1097]: return errors.Trace(err2)

**Review by tiancaiamao — APPROVED** [review]

**Comment by morgo** [comment]
@dveeden Can you please update https://docs.pingcap.com/tidb/dev/sql-statement-load-data#mysql-compatibility and https://docs.pingcap.com/tidb/dev/mysql-compatibility#unsupported-features

**Pull request merged**